### PR TITLE
Add support for testing installs from mirrored repositories

### PR DESCRIPTION
**Pull request by umohnani8**
Taking over https://github.com/openshift/release/pull/4472

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>
Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

### CODE REVIEW

**Comment by umohnani8** [comment]
@smarterclayton PTAL, seems like this is what we need for the periodic job, but I might be missing some things. Should the PR run the periodic job I just added, I don't see it in the list of jobs it is running?

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml [hunk 1]
```diff
@@ -708,9 +714,28 @@ objects:
         if [[ -n "${INSTALL_INITIAL_RELEASE}" && -n "${RELEASE_IMAGE_INITIAL}" ]]; then
           echo "Installing from initial release ${RELEASE_IMAGE_INITIAL}"
           OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="${RELEASE_IMAGE_INITIAL}"
+        elif [[ "${TEST_ENV_MIRROR}" == "true" && "${JOB_NAME_SAFE}" == "e2e-aws" ]]; then
```
Comment (smarterclayton) [inline ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: remove the JOB_NAME_SAFE check

File: ci-operator/jobs/openshift/release/openshift-release-release-4.2-periodics.yaml [hunk 2]
```diff
@@ -1785,3 +1785,80 @@ periodics:
     - name: pull-secret
       secret:
         secretName: ci-pull-credentials
+- agent: kubernetes
+  cron: ""
+  decorate: true
+  interval: 12h
+  labels:
+    pj-rehearse.openshift.io/can-be-rehearsed: "true"
+  name: release-openshift-origin-installer-e2e-aws-install-mirrors-4.2
```
Comment (abhinavdahiya) [inline ci-operator/jobs/openshift/release/openshift-release-release-4.2-periodics.yaml]: this should be canary like others...

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml [hunk 3]
```diff
@@ -25,6 +25,7 @@ parameters:
 - name: ENABLE_PROXY
 - name: BUILD_ID
   required: false
+- name: ENABLE_MIRROR_TEST
```
Comment (abhinavdahiya) [inline ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: nit: `ENABLE_MIRROR` should be enough

**Comment by umohnani8** [comment]
currently blocked on https://bugzilla.redhat.com/show_bug.cgi?id=1741391

**Comment by cuppett** [comment]
cc @wking

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml (hunk 3 above)
Comment (wking) [reply ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: > nit: `ENABLE_MIRROR` should be enough

Fixed in ccbc46931 -> 8843aa174.

File: ci-operator/jobs/openshift/release/openshift-release-release-4.2-periodics.yaml (hunk 2 above)
Comment (wking) [reply ci-operator/jobs/openshift/release/openshift-release-release-4.2-periodics.yaml]: > this should be canary like others...

Fixed in ccbc469 -> 8843aa1.

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml [hunk 4]
```diff
@@ -710,9 +725,29 @@ objects:
         if [[ -n "${INSTALL_INITIAL_RELEASE}" && -n "${RELEASE_IMAGE_INITIAL}" ]]; then
           echo "Installing from initial release ${RELEASE_IMAGE_INITIAL}"
           OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="${RELEASE_IMAGE_INITIAL}"
+        elif [[ "${ENABLE_MIRROR_TEST}" == "true" ]]; then
+          export PATH=$PATH:/tmp/bin
+          mkdir /tmp/bin
+          curl https://storage.googleapis.com/origin-ci-test/_tools/4.1.0/oc >/tmp/bin/oc 2>/dev/null
```
[the remark this answers is not in the document]
Comment (wking) [reply ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: > I think we could just get the of binary from the test container

Fixed in ccbc469 -> 8843aa1.

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml (hunk 1 above)
Comment (wking) [reply ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: > remove the JOB_NAME_SAFE check

Fixed at some point before 53490e3.

**Comment by wking** [comment]
[Load-balancer timeout][1] is [rhbz#1717604][2].

/retest

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/4585/rehearse-4585-canary-openshift-ocp-installer-e2e-aws-mirrors-4.2/22
[2]: https://bugzilla.redhat.com/show_bug.cgi?id=1717604

**Comment by wking** [comment]
/retest

now that `pj-rehearse` has wrapped up.

**Comment by wking** [comment]
[Mirror failure][1]:

```
Success
Update image: registry.svc.ci.openshift.org/ci-op-3c5fkrpt/mirrored:release
Mirror prefix: registry.svc.ci.openshift.org/ci-op-3c5fkrpt/mirrored
info: Mirroring completed in 2m59.44s (27.82MB/s)
info: Pushed to registry.svc.ci.openshift.org/ci-op-3c5fkrpt/mirrored:release
sha256:a065d27010e8986bd83b9bd59d5368ac7e29d71545efca4a950ca469c9deacf0 4.2.0-0.nightly-2019-09-18-194238 2019-09-18T19:47:07Z
error: the release could not be reproduced from its inputs
```

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/4585/rehearse-4585-canary-openshift-ocp-installer-e2e-aws-mirrors-4.2/23

**Comment by wking** [comment]
[Failed][1] (in `oc adm release new ...`?) with:

```
info: Pushed to registry.svc.ci.openshift.org/ci-op-52kplskn/mirrored:release
sha256:2a60a7a22eea04f9a195cf6f45950d86b8e1442b25049f703b837b8b32f6fd3c 4.2.0-0.nightly-2019-09-18-211009 2019-09-18T21:12:13Z
error: the release could not be reproduced from its inputs
```

You may want a `set -x` like [this][2] to help with debugging `new` vs. `mirror`.

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_release/4585/rehearse-4585-canary-openshift-ocp-installer-e2e-aws-mirrors-4.2/28/build-log.txt
[2]: https://github.com/openshift/release/pull/5041/commits/ca7fad8e5fc2af0cd48ed52248c9e23b65d22f91#diff-e9ff14653190666556f490397d19c553R790

**Comment by wking** [comment]
[Failed][1] with:

```
error: the release could not be reproduced from its inputs
```

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/4585/rehearse-4585-canary-openshift-ocp-installer-e2e-aws-mirrors-4.2/29

**Comment by umohnani8** [comment]
@wking yeah I forgot to actually commit the changes for picking up the newer oc, my bad. Should hopefully pass this stage now

**Comment by wking** [comment]
I'm [getting the same error][1] with the from-`test` `oc`.  I'll try to figure out what it means when I get in ;).

[1]: https://github.com/openshift/release/pull/5041#issuecomment-533142769

**Comment by umohnani8** [comment]
@wking yup, just failed for me with the same error again...

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml [hunk 5]
```diff
@@ -925,6 +1034,8 @@ objects:
         value: /etc/openshift-installer/gce.json
       - name: KUBECONFIG
         value: /tmp/artifacts/installer/auth/kubeconfig
+      - name: ENABLE_MIRROR
+        value: ${ENABLE_MIRROR}
```
Comment (wking) [inline ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: nit: we don't actually need this in `teardown`, do we?

**Comment by wking** [comment]
/retitle Add support for testing installs from mirrored repositories

Pull the WIP, because this is almost good to go :+1:

File: ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml (hunk 5 above)
Comment (umohnani8) [reply ci-operator/templates/openshift/installer/cluster-launch-installer-e2e.yaml]: We don't, fixed.

**Comment by umohnani8** [comment]
Tests are green! This is ready for review and merge. @wking @smarterclayton @abhinavdahiya PTAL

**Pull request merged**